### ISSUE TITLE
inconsistency in mark read and message list

### ISSUE THREAD

**Issue by pcause** [martinrotter/rssguard#243]
On Version: 3.6.3 (built on Windows/x86_64)
Revision: daad076

I have show only unread enabled.

If i select a message it goes from bold to normal meaning it is marked as read bit is still displayed.  if I select an item and then hit a key stroke to open in external browser the item is gone from the message list.  I don't think this should happen as I lose the ability to mark unread or perform any other actions.  All items should stay in the list until I leave the selected feed/categiry

**Comment by martinrotter** [comment]
Yes, this is rather big limitation in current implementation, because in some scenarios, the list is in fact completely reloaded from the database, which is sometimes fast situation than somehow cache some past changes and reload only individual messages.

I will think about it, but maybe we will have to sacrifice some performance for the better UX.

**Comment by pcause** [comment]
if there is an update in progress, once the update is done the feed/category displayed is updated and you loose read items.  perhaps remember last fetched time from db for what is displayed and then fetch new messages from the database and add to the list without complete data refresh.??

**Comment by martinrotter** [comment]
> if there is an update in progress, once the update is done the feed/category displayed is updated and you loose read items. perhaps remember last fetched time from db for what is displayed and then fetch new messages from the database and add to the list without complete data refresh.??

Yes. When update of feeds is finished, then the list of messages is repopulated from the database and all filters (regexp, unread) are reapplied. This is because when feeds get update, it might actually happen the the message you are viewing at that moment is not up-to-date one because newer/updated message got downloaded and stored in DB, thus you would see and work with old message which could lead to some confusions.

Moreover, in TT-RSS the message could even be marked as deleted in TT-RSS server and when feed update is finished, you would work with actually deleted message which, again, could result in "weird" message states.

But I can agree that advanced user who really knows what he is doing could perhaps benefit from NOT having the message list forcefully reloaded once feeds are updated.

Perhaps create a separate issue ticket for this specific stuff.

**Comment by martinrotter** [comment]
OK, so I commited 0eb8d756.

Can you please test if message list is now more consistent when dev build gets compiled?

**Comment by pcause** [comment]
Just tried opening in external browser.  The item is marked as read but not removed from the display.  Will do some more testing during the day and share results here.

**Comment by pcause** [comment]
one other thing I just did/noticed.  If I select an intem in the message list and hit the down arrow, each subsequent item is selected and marked read.  If I hit page down, you go a full scrren and select that item, but the above items are left unread.  This is inconsistent.  First, I would try to scroll 1/2 scren so I have context and I would put the selected line in the middle of the screen and mark aqll above it to where I was when I hit page down as read.   Page up should also go 1/2 screen but question of whether you also mark read intervening items.  Consistency would say yes.  perhaps shift+page {up/down} moves the selected item without marking intervening read.

**Comment by martinrotter** [comment]
No, it is not inconsistent, because purpose of Pg UP/DOWN buttons is to **jump** over several items and move fast in the list. If you jump in the list it does not mean that you should mark all jumped-over-items as read.

For example Mozilla Thunderbird (e-mail client) behaves in this regard the same way as RSS Guard, as well as vast majority of similar applications which have "list of messages".

Having a separate menu item/action to mark all above/below messages as read/unread is fine and I can add it, but hijacking Pg UP/DOWN for kind of similar action is no-go for me.

You can keep selection in the middle of msg list viewport, this feature is already there.

![image](https://user-images.githubusercontent.com/1255302/84893681-e797df00-b09f-11ea-9eef-ef509e63515b.png)

The number of skipped items in Pg UP/DOWN feature is hardcoded in Qt library and it seems to jump over 12 items if possible, which is quite big number. I would love to jump it over like maybe 6 items.

The only difference I have found is that when you do extended selection via arrows+SHIFT buttons, then all messages are marked read (as you gradually select them). Thunderbird does not do that and we can certainly talk about this behavior, although I find it rather useful. For this last behavior, feel free to open bug ticket.

As for the behavior related to "show unread messages only" is it all now working? Ie. no random disappearing of items in the middle of browsing?

**Comment by pcause** [comment]
OK, thanks.  Firs the disappearing seems gone.

The "keep selection" wasn't clear as to purpose and now I will check and use.

Understand about pgup/down.  perhaps have this be shift pg up/down?? or maybe just allow to bind to a key?

**Comment by martinrotter** [comment]
Yes, "perhaps have this be shift pg up/down". Shift+Pg UP or DOWN is expected to select some number of consecutive items in a list while last item is also displayed. Without shift, no selections are done and we jump across several items.

I see, that you are concerned that in some scenarios some messages are marked as read and other scenarios, they are not (when multiple messages are selected). This ticket was about disappearing items, this is hopefully solved.

If you want me to unify behavior when multiple messages are selected, create separate ticket pls. :)
